packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_def.hpp

```diff
@@ -2835,91 +2925,92 @@ class GetLocalDiagCopy {
-      // latter, we replace it with zero (that row is not owned by the
-      // calling process, so it has no entries to pack).
-      if (numEnt == Teuchos::OrdinalTraits<size_t>::invalid ()) {
-        numEnt = 0;
-      }
-      const size_t numScalarEnt = numEnt * blockSize * blockSize;
 
-      // The 'if' branch implicitly assumes that packRowCount() returns
-      // zero if numEnt == 0.
-      size_t numBytesPerValue = 0;
-      if (numEnt > 0) {
-        // Get a locally indexed view of the current row's data.  If
-        // the current row has > 0 entries, we need an entry in order
-        // to figure out the byte count of the packed row.  (We really
-        // only need it if ST's size is determined at run time.)
-        Scalar* valsRaw = NULL;
-        const LO* lidsRaw = NULL;
-        LO actualNumEnt = 0;
-        const LO errCode =
-          src->getLocalRowView (lclRow, lidsRaw, valsRaw, actualNumEnt);
-
-        if (numEnt != static_cast<size_t> (actualNumEnt)) {
-          std::ostream& err = this->markLocalErrorAndGetStream ();
-          err << "packAndPrepare: Local row " << i << " claims to have " <<
-            numEnt << "entry/ies, but the View returned by getLocalRowView() "
-            "has " << actualNumEnt << " entry/ies.  This should never happen.  "
-            "Please report this bug to the Tpetra developers." << endl;
-          return;
-        }
-        if (errCode == Teuchos::OrdinalTraits<LO>::invalid ()) {
-          std::ostream& err = this->markLocalErrorAndGetStream ();
-          err << "packAndPrepare: Local row " << i << " is not in the row Map "
-            "of the source object on the calling process." << endl;
-          return;
-        }
+    Impl::BlockCrsRowStruct<size_t> rowReducerStruct;
 
-        const ST* valsRawST =
-          const_cast<const ST*> (reinterpret_cast<ST*> (valsRaw));
-        View<const ST*, HES, MemoryUnmanaged> vals (valsRawST, numScalarEnt);
+    // Graph information is on host; let's do this on host parallel reduce
+    // Sync necessary data to host
 
-        // NOTE (mfh 07 Feb 2015) Since we're using the host memory
-        // space here for now, this doesn't assume UVM.  That may change
-        // in the future, if we ever start packing on the device.
-        numBytesPerValue = PackTraits<ST, HES>::packValueCount (vals(0));
+    if (exportLIDs.need_sync_host()) {
+      Kokkos::DualView<const local_ordinal_type*, device_type> exportLIDsTemp = exportLIDs;
+      exportLIDsTemp.sync_host();
+    }
+    auto exportLIDsHost = exportLIDs.view_host();
+    auto numPacketsPerLIDHost = numPacketsPerLID.view_host(); // we will modify this
+    {
+      typedef Impl::BlockCrsReducer<Impl::BlockCrsRowStruct<size_t>,host_exec> reducer_type;
+      const auto policy = Kokkos::RangePolicy<host_exec>(0, numExportLIDs);
+      Kokkos::parallel_reduce
+        (policy, 
+         [=](const int &i, typename reducer_type::value_type &update) {
```

Comment:
We should actually label these loops so that Kokkos Profiling can pick them up.  I'm OK with that waiting for the next PR, though.